File: src/helper/filesystem/search_file_directory.c
/**
* search_file_directory.c
*
* (C) Copyright 2025 AtlantisOS Project
* by @NachtsternBuild
*
* License: GNU GENERAL PUBLIC LICENSE Version 3
*
*
* @brief Search if a predefined string exsists in the file of predefined directory
*
* Usage:
*  int result = search_file_directory(SOURCES_DIR, SEARCH_STRING); 
*  if (result == -1) {
*  	  return 2; // error
*  } else if (result == 1) {
*     return 0; // no file
*  } else {
*     return 1; // file
*  }
*
* Notes: 
* @param directory: Search at this path.
* @param search_string: Search for this string.
* @return 0 file not exsists, 1 file exsists, -1 on error, .
*/

#include "helper.h"

int search_file_directory(const char *directory, const char *search_string) 
{
    DIR *dir;
    struct dirent *entry;
    int found = 0;

    dir = opendir(directory);
    if (!dir) 
    {
        perror("Kann Verzeichnis nicht öffnen");
        return -1; 
    }

    while ((entry = readdir(dir)) != NULL) 
    {
        // ignore . and ..
        if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0)
        {
            continue;
		}
		
		// found file/files with string
        if (strstr(entry->d_name, search_string) != NULL) 
        {
            found = 1;
            break; 
        }
    }

    closedir(dir);
    return found;
}

File: src/helper/filesystem/search_file_directory.rs
//! Search for filenames within a directory.
//!
//! This module searches the filenames in a specified directory for 
//! a substring. It is useful for checking whether files exist 
//! whose full names are not known exactly.
/**
* search_file_directory.c
*
* (C) Copyright 2026 AtlantisOS Project
* by @NachtsternBuild
*
* License: GNU GENERAL PUBLIC LICENSE Version 3
*/

use std::fs;
use std::path::Path;

/// Checks whether a search term appears in the name of a file within a directory.
///
/// The function scans the directory flat (not recursively) and compares 
/// each filename with the passed `search_string`.
///
/// # Arguments
/// * `directory` - The path to the directory to be searched (accepts `&str`, `PathBuf`, etc.).
/// * `search_string` - The string to search for in the filename.
///
/// # Return Value
/// * `Ok(true)` - At least one file contains the search term in its name.
/// * `Ok(false)` - No file matches the search criteria.
/// * `Err(std::io::Error)` - Error accessing the directory.
///
/// # Usage:
/// 
/// ```rust
/// fn main() {
///    let sources_dir = "./src";
///    let search_string = "helper";
///    match search_file_directory(sources_dir, search_string) {
///        Ok(true) => println!("File!"),
///        Ok(false) => println!("No file."),
///        Err(e) => eprintln!("Error: {}", e),
///    }
/// }
/// ```
 
pub fn search_file_directory<P: AsRef<Path>>(directory: P, search_string: &str) -> Result<bool, std::io::Error> {
    let entries = fs::read_dir(directory)?;

    for entry in entries {
        let entry = entry?;
        let file_name = entry.file_name();
        
        // checks whether the filename contains the search string
        // convert OsString to a normal string for comparison
        if let Some(name_str) = file_name.to_str() {
            if name_str.contains(search_string) {
                return Ok(true);
            }
        }
    }

    Ok(false)
}
